### PR TITLE
Typing: Narrow down types of arguments (NDFrame)

### PR DESCRIPTION
**Pull request by kevinanker**
- [x] closes parts of # 10 (in https://github.com/noatamir/pydata-workshop)
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.

Updated item 11-14 (15 seemed already done):
 - [x] Specify method in to_sql more precisely
 - [x] Specify errors in to_csv more precisely
 - [x] Specify mode in to_csv more precisely
 - [x] Specify kind and na_position in sort_values more precisely

### CODE REVIEW

**Comment by noatamir** [comment]
Thanks for the PR @kjanker! There are a couple of errors identified by pre-commit, which I think you can solve yourself. Let me know if you need help understanding the errors.

File: pandas/core/generic.py [hunk 1]
```diff
@@ -3548,7 +3549,7 @@ def to_csv(
         header: bool_t | list[str] = ...,
         index: bool_t = ...,
         index_label: IndexLabel | None = ...,
-        mode: str = ...,
+        mode: Literal["w", "x", "a"] = ...,
```
Comment (twoertwein) [inline pandas/core/generic.py:3552]: I think many more are allowed for `mode`, `"wt", "wb", ...` not sure whether it is worth narrowing it.

Comment (kjanker) [reply pandas/core/generic.py:3552]: I think you are right. One could add all options from build-in ``open``, but maybe it's not worth it.

Character | Meaning
-- | --
'r' | open for reading (default)
'w' | open for writing, truncating the file first
'x' | open for exclusive creation, failing if the file already exists
'a' | open for writing, appending to the end of file if it exists
'b' | binary mode
't' | text mode (default)
'+' | open for updating (reading and writing)

Comment (twoertwein) [reply pandas/core/generic.py:3552]: Typeshed enumerated all of the combinations. https://github.com/python/typeshed/blob/main/stdlib/_typeshed/__init__.pyi#L139

We might be able to import them inside a TYPE_CHECKING block and then use that in the annotations:

` mode: _typeshed.OpenTextModeUpdating | _typeshed.OpenBinaryModeUpdating = ...,`

(Not sure whether that is a 'good' suggestions)

**Review by twoertwein — APPROVED** [review]

**Comment by twoertwein** [comment]
Thanks @kjanker!

**Pull request merged**